fuel-metrics/src/p2p_metrics.rs

```diff
@@ -0,0 +1,48 @@
+use lazy_static::lazy_static;
+use once_cell::race::OnceBox;
+use prometheus_client::{
+    metrics::counter::Counter,
+    registry::Registry,
+};
+use std::{
+    boxed::Box,
+    default::Default,
+};
+
+pub struct P2PMetrics {
+    pub gossip_sub_registry: OnceBox<Registry>,
```

Comment:
I'm not sure it is the right pattern to have a separate `Registry` inside each `{}Metrics` structure. It looks redundant because the `{}Metrics` already contains all fields.

Maybe it is better to have one `Registry` that is used by each service. This service can be used to encode all metrics at once.

In this case, we don't need many static metrics for each service, we have one. For each service(crate), we can define its metric inside and add the `init` method and `getters` via traits.

Like:

```rust
pub trait TXPoolMetrics {
    fn init(&mut self); // Maybe return an error, like duplicate in registry?

    fn tx_pool(&self) -> TxPool;
}
```

We don't need to pass metrics into each service constructor in this case. Because the register is already static, you need to call `init` only if the `metrics` feature is enabled. Also, you don't need a workaround with `gossip_sub_registry`.

Reply:
@FuelLabs/client What do you think about the idea with one global `Registry`?

Reply:
This won't work because you can't have `&mut self` in a global without using locks.

Reply:
We need `&mut self` only during initialization once so `RefCell` works for us=)

Reply:
Do you mean `OnceCell`? I don't think a `RefCell` can be used in a global because they aren't `Sync`.

Reply:
Oh, wait, we don't need to use any cell, we need call `init` during `lazy_static`=)

Reply:
I've removed `init` and added it to `::new` method

Reply:
I meant for the idea with traits, we can use the `init` method in the `lazy_static` to init metrics for each service. In this case, we don't need to worry about `&mut self` because it is used only during lazy initialization.

Reply:
@ControlCplusControlV didn't you need the `&mut self` to pass to libp2p?

Reply:
Yes the `&mut self` is needed to initialize the gossipsub with metrics

Reply:
If you initialize all metrics in one `lazy_static` call, you don't need `&mut self` outside of that lazy initialization

Reply:
I've generally done that where possible, however `gossip_sub` metrics require the `&mut self` to be passed across into `fuel-p2p` as documented [here](https://docs.rs/libp2p-gossipsub/latest/libp2p_gossipsub/struct.Gossipsub.html#method.new_with_metrics) with the initialization of the p2p gossipsub service

Reply:
Oh, I see. We need it here=)
<img width="709" alt="image" src="https://user-images.githubusercontent.com/18346821/200089675-8f684fe9-d9fe-4f94-bd8d-2b2d92f3940c.png">

Then yeah, `OnceCell` can be our choice=)

Reply:
Global metrics are required because it is the lesser of evils when it comes to plumbing metrics. To pass instances of counters and registries would be nightmarish in the amount of functions and service handlers it would touch, especially if this gets abstracted to a compile time flag later.

Since without global metrics we end up with scenarios where `Counters` need to be plumbed all the way to down to the basic `write` and `get` calls of the Database (which would pollute a lot of function signatures)